code/obj/item/pet_carrier.dm

```diff
@@ -0,0 +1,331 @@
+		src.vis_contents.Add(src.vis_contents_proxy)
+
+		// Instantiate the grate.
+		src.grate_proxy = new()
+		src.grate_proxy.icon = src.icon
+		src.grate_proxy.vis_flags |= src.carrier_vis_flags
+		src.vis_contents.Add(src.grate_proxy)
+
+		src.UpdateIcon()
+
+		if (src.default_mob_type)
+			if (!ispath(src.default_mob_type, src.allowed_mob_type))
+				return
+			var/mob/spawned_mob = new src.default_mob_type
+			if (spawned_mob)
+				src.add_mob(spawned_mob)
+
+	disposing()
+		for (var/occupant in src.carrier_occupants)
+			src.eject_mob(occupant)
+		src.vis_contents = null
+		qdel(src.grate_proxy)
+		src.grate_proxy = null
+		qdel(src.vis_contents_proxy)
+		src.vis_contents_proxy = null
+		..()
+
+	examine()
+		. = ..()
+		var/carrier_occupants_length = length(src.carrier_occupants)
+		if (carrier_occupants_length)
+			if (carrier_occupants_length > src.explicit_name_limit)
+				. += "There's a whole zoo inside!"
+				return
+			. += "It's carrying [english_list(src.carrier_occupants)]."
+
+	update_icon()
+		..()
+		if (length(src.carrier_occupants))
+			src.grate_proxy.icon_state = src.grate_closed_icon_state
+			src.item_state = src.carrier_closed_item_state
+		else
+			src.grate_proxy.icon_state = src.grate_open_icon_state
+			src.item_state = src.carrier_open_item_state
+
+	attack(mob/M, mob/user)
+		if (user.a_intent == INTENT_HARM)
+			return ..()
+		if (istype(M))
+			if (!src.find_empty_hand(user))
+				boutput(user, "<span class='alert'>You need a free hand to scoop up [M]!</span>")
+				return ..()
+			if (!istype(M,src.allowed_mob_type))
+				boutput(user, "<span class='alert'>[M] can't quite fit inside [src]!</span>")
+				return ..()
+			if (src.carrier_max_capacity <= length(src.carrier_occupants))
+				boutput(user, "<span class='alert'>[src] is too crowded to fit one more!</span>")
+				return ..()
+			actions.start(new /datum/action/bar/icon/pet_carrier(M, src, src.icon, src.trap_mob_icon_state, TRAP_MOB, src.actionbar_duration), user)
+			return
+		..()
+
+	attack_self(mob/user)
+		src.attempt_removal(user)
+		..()
+
+	// Ensure that things inside can actually breathe.
+	remove_air(amount)
+		var/turf/current_turf = get_turf(src)
+		. = current_turf.remove_air(amount)
+
+	return_air()
+		var/turf/current_turf = get_turf(src)
+		. = current_turf.return_air()
+
+	mob_flip_inside(mob/user)
+		..(user)
+		src.mob_resist_inside(user)
+
+	mob_resist_inside(mob/user)
+		if (ON_COOLDOWN(src, "resist_damage", 3 SECONDS))
+			return
+		animate_storage_thump(src)
+		if (!src.can_break_out)
+			boutput(user, "<span class='alert'>It's no use! You can't leave [src]!</span>")
+			return
+		boutput(user, "<span class='alert'>You try to bust open the door of [src]!</span>")
+		src.take_door_damage(src.damage_per_resist)
+
+	throw_impact(atom/hit_atom, datum/thrown_thing/thr)
+		..()
+		if (src.carrier_occupants)
+			animate_storage_thump(src)
+		for (var/mob/occupant in src.carrier_occupants)
+			occupant.throw_impact(hit_atom, thr)
```

Comment:
This janks out with disposal chutes, since those move you inside them when you impact them- phases the animals out of the carrier into the chute and doesn't update the carrier properly.
![image](https://github.com/goonstation/goonstation/assets/67879783/906c3d64-f807-4d2d-9a57-2ce7b274f333)

Cleanest way to fix (along with other critter-teleported-out bugs) is to add an `Exited()` override that checks if the exiting thing is in `carrier_occupants`, and then calls `eject_mob` on it if it is. (Won't cause any `loc` fuckery since you used `MOVE_OUT_TO_TURF_SAFE` >:) )